sys/shell/shell.c

```diff
@@ -45,171 +46,161 @@ static void _putchar(int c) {
 {
-    printf("%-20s %s\n", "Command", "Description");
-    puts("---------------------------------------");
-
-    const shell_command_t *command_lists[] = {
-        command_list,
-#ifdef MODULE_SHELL_COMMANDS
-        _shell_command_list,
-#endif
-    };
+    for (; entry->name != NULL; entry++) {
+        printf("%-20s %s\n", entry->name, entry->desc);
+    }
+}
 
-    /* iterating over command_lists */
-    for (unsigned int i = 0; i < ARRAY_SIZE(command_lists); i++) {
+static void print_help(const shell_command_t *command_list)
+{
+    puts("Command              Description\n---------------------------------------");
 
-        const shell_command_t *entry;
+    if (command_list != NULL) {
+        print_commands(command_list);
+    }
 
-        if ((entry = command_lists[i])) {
-            /* iterating over commands in command_lists entry */
-            while (entry->name != NULL) {
-                printf("%-20s %s\n", entry->name, entry->desc);
-                entry++;
-            }
-        }
+    if (_builtin_cmds != NULL) {
+        print_commands(_builtin_cmds);
     }
 }
 
+enum PARSE_STATE {
+    PARSE_SPACE,
+    PARSE_UNQUOTED,
+    PARSE_SINGLEQUOTE,
+    PARSE_DOUBLEQUOTE,
+    PARSE_ESCAPE_MASK,
+    PARSE_UNQUOTED_ESC,
+    PARSE_SINGLEQUOTE_ESC,
+    PARSE_DOUBLEQUOTE_ESC,
+};
+
+static enum PARSE_STATE escape_toggle(enum PARSE_STATE s)
+{
+    return s ^ PARSE_ESCAPE_MASK;
+}
+
+#define SQUOTE '\''
+#define DQUOTE '"'
+#define ESCAPECHAR '\\'
+#define BLANK ' '
+
 static void handle_input_line(const shell_command_t *command_list, char *line)
 {
     static const char *INCORRECT_QUOTING = "shell: incorrect quoting";
 
     /* first we need to calculate the number of arguments */
-    unsigned argc = 0;
-    char *pos = line;
-    int contains_esc_seq = 0;
-    while (1) {
-        if ((unsigned char) *pos > ' ') {
-            /* found an argument */
-            if (*pos == '"' || *pos == '\'') {
-                /* it's a quoted argument */
-                const char quote_char = *pos;
-                do {
-                    ++pos;
-                    if (!*pos) {
-                        puts(INCORRECT_QUOTING);
-                        return;
-                    }
-                    else if (*pos == '\\') {
-                        /* skip over the next character */
-                        ++contains_esc_seq;
-                        ++pos;
-                        if (!*pos) {
-                            puts(INCORRECT_QUOTING);
-                            return;
-                        }
-                        continue;
-                    }
-                } while (*pos != quote_char);
-                if ((unsigned char) pos[1] > ' ') {
-                    puts(INCORRECT_QUOTING);
-                    return;
+    int argc = 0;
+    char *readpos = line;
+    char *writepos = readpos;
+    enum PARSE_STATE pstate = PARSE_SPACE;
+
+    while (*readpos != '\0') {
```

Comment:
I find this part too hard to read. Maybe something like?

```suggestion
    while (*readpos != '\0') {
        char wordbreak;

        switch (pstate) {
            case PARSE_SPACE:
                if (*readpos != BLANK) {
                    argc++;
                }
                if (*readpos == SQUOTE) {
                    pstate = PARSE_SINGLEQUOTE;
                } else if (*readpos == DQUOTE) {
                    pstate = PARSE_DOUBLEQUOTE;
                } else if (*readpos == ESCAPECHAR) {
                    pstate = PARSE_UNQUOTED_ESC;
                } else if (*readpos != BLANK) {
                    pstate = PARSE_UNQUOTED;
                    goto add;
                }
                readpos++;
                continue;
            case PARSE_UNQUOTED:
                wordbreak = BLANK;
                goto break_word;
            case PARSE_SINGLEQUOTE:
                wordbreak = SQUOTE;
                goto break_word;
            case PARSE_DOUBLEQUOTE:
                wordbreak = DQUOTE;
                goto break_word;
            default: /* QUOTED state */
                pstate = escape_toggle(pstate);
                goto add;
        }

break_word:
    if (*readpos == wordbreak) {
        pstate = PARSE_SPACE;
        *writepos++ = '\0';
        goto parse_end;
    } else if (*readpos == ESCAPECHAR) {
        pstate = escape_toggle(pstate);
        goto parse_end;
    }

add:
        *writepos++ = *readpos;

parse_end:
        readpos++;
    }
```
It's a similar `goto` logic but the labels are not inside the switch.
Also, it seems to reduce the size from:
```
Building application "default" for "native" with MCU "native".
   text    data     bss     dec     hex filename
  90573    1104   72088  163765   27fb5 /home/jialamos/Development/RIOT/examples/default/bin/native/default.elf
```
to
```
Building application "default" for "native" with MCU "native".

   text    data     bss     dec     hex filename
  90557    1104   72088  163749   27fa5 /home/jialamos/Development/RIOT/examples/default/bin/native/default.elf

```

On samr21-xpro, the build size keeps the same.

Reply:
Here's a non-goto version:
```suggestion
    for(readpos = line; *readpos != '\0'; readpos++) {
        char wordbreak;
        switch (pstate) {
            case PARSE_SPACE:
                if (*readpos != BLANK) {
                    argc++;
                }
                if (*readpos == SQUOTE) {
                    pstate = PARSE_SINGLEQUOTE;
                } else if (*readpos == DQUOTE) {
                    pstate = PARSE_DOUBLEQUOTE;
                } else if (*readpos == ESCAPECHAR) {
                    pstate = PARSE_UNQUOTED_ESC;
                } else if (*readpos != BLANK) {
                    pstate = PARSE_UNQUOTED;
                    *writepos++ = *readpos;
                }
                continue;
            case PARSE_UNQUOTED:
                wordbreak = BLANK;
                break;
            case PARSE_SINGLEQUOTE:
                wordbreak = SQUOTE;
                break;
            case PARSE_DOUBLEQUOTE:
                wordbreak = DQUOTE;
                break;
            default: /* QUOTED state */
                pstate = escape_toggle(pstate);
                *writepos++ = *readpos;
                continue;
        }

        if (*readpos == wordbreak) {
            pstate = PARSE_SPACE;
            *writepos++ = '\0';
        } else if (*readpos == ESCAPECHAR) {
            pstate = escape_toggle(pstate);
        }

    }
```

Reply:
I made a similar modification (I'm about to push the fixup). The size does not change in samr21 because the compiler is smart enough to move the code around, so the generated machine code is the same in both cases.

Reply:
mmm, hold on, I don't think it improves so much in terms of clarity:
```c
    while (*readpos != '\0') {
        char wordbreak;

        switch (pstate) {
            case PARSE_SPACE:
                if (*readpos != BLANK) {
                    argc++;
                }
                if (*readpos == SQUOTE) {
                    pstate = PARSE_SINGLEQUOTE;
                } else if (*readpos == DQUOTE) {
                    pstate = PARSE_DOUBLEQUOTE;
                } else if (*readpos == ESCAPECHAR) {
                    pstate = PARSE_UNQUOTED_ESC;
                } else if (*readpos != BLANK) {
                    pstate = PARSE_UNQUOTED;
                    goto write_word;
                }
                goto parse_end;
            case PARSE_UNQUOTED:
                wordbreak = BLANK;
                break;
            case PARSE_SINGLEQUOTE:
                wordbreak = SQUOTE;
                break;
            case PARSE_DOUBLEQUOTE:
                wordbreak = DQUOTE;
                break;
            default: /* QUOTED state */
                pstate = escape_toggle(pstate);
                goto write_word;
        }

        if (*readpos == wordbreak) {
            pstate = PARSE_SPACE;
            *writepos++ = '\0';
        } else if (*readpos == ESCAPECHAR) {
            pstate = escape_toggle(pstate);
        } else {
write_word:
            *writepos++ = *readpos;
        }
parse_end:
        readpos++;
    }
```

Reply:
I could draw the state diagram as a comment before the function.

Reply:
the state diagram would also help